lib/SILOptimizer/Mandatory/Differentiation.cpp

```diff
@@ -2117,17 +2117,9 @@ void DifferentiableActivityInfo::setUsefulAcrossArrayInitialization(
       auto *ptai = dyn_cast<PointerToAddressInst>(use->getUser());
       assert(ptai && "Expected `pointer_to_address` user for uninitialized "
                      "array intrinsic");
-      for (auto use : ptai->getUses()) {
-        auto *inst = use->getUser();
-        if (auto *si = dyn_cast<StoreInst>(inst)) {
-          setUsefulAndPropagateToOperands(si->getSrc(), dependentVariableIndex);
-        } else if (auto *iai = dyn_cast<IndexAddrInst>(inst)) {
-          for (auto use : iai->getUses())
-            if (auto si = dyn_cast<StoreInst>(use->getUser()))
-              setUsefulAndPropagateToOperands(si->getSrc(),
-                                              dependentVariableIndex);
-        }
-      }
+      // Propagate usefulness through `pointer_to_address` users' operands.
+      for (auto use : ptai->getUses())
+        propagateUseful(use->getUser(), dependentVariableIndex);
```

Comment:
Rather than calling `propagateUseful`, I also tried adding more ad-hoc logic for propagating usefulness to `pointer_to_address` users:

```cpp
      // Propagate usefulness through `pointer_to_address` users' operands.
      for (auto use : ptai->getUses()) {
        auto *inst = use->getUser();
        if (auto *si = dyn_cast<StoreInst>(inst)) {
          setUsefulAndPropagateToOperands(si->getSrc(), dependentVariableIndex);
        } else if (auto *cai = dyn_cast<CopyAddrInst>(inst)) {
          setUsefulAndPropagateToOperands(cai->getSrc(),
                                          dependentVariableIndex);
        } else if (auto *iai = dyn_cast<IndexAddrInst>(inst)) {
          for (auto use : iai->getUses()) {
            if (auto si = dyn_cast<StoreInst>(use->getUser())) {
              setUsefulAndPropagateToOperands(si->getSrc(),
                                              dependentVariableIndex);
            } else if (auto *cai = dyn_cast<CopyAddrInst>(inst)) {
              setUsefulAndPropagateToOperands(cai->getSrc(),
                                              dependentVariableIndex);
            }
          }
        }
      }
```

However, this duplicates code from `propagateUseful` and seems less maintainable. Activity analysis also seems not quite right:
```swift
@differentiable(where T: Differentiable)
func testArrayUninitializedIntrinsicGeneric<T>(_ x: T, _ y: T) -> [T] {
  return [x, y]
}
```
```
[AD] Activity info for $s17activity_analysis38testArrayUninitializedIntrinsicGenericySayx
Gx_xtlF at (source=0 parameters=(0 1))
bb0:
[ACTIVE] %0 = argument of bb0 : $*T                        // users: %10, %2
[VARIED] %1 = argument of bb0 : $*T                        // users: %13, %3
[USEFUL]   %4 = integer_literal $Builtin.Word, 2           // user: %6
[NONE]   // function_ref _allocateUninitializedArray<A>(_:)
  %5 = function_ref @$ss27_allocateUninitializedArrayySayxG_BptBwlF : $@convention(thin)
<τ_0_0> (Builtin.Word) -> (@owned Array<τ_0_0>, Builtin.RawPointer) // user: %6
[ACTIVE]   %6 = apply %5<T>(%4) : $@convention(thin) <τ_0_0> (Builtin.Word) -> (@owned Ar
ray<τ_0_0>, Builtin.RawPointer) // user: %7
[ACTIVE] (**%7**, %8) = destructure_tuple %6 : $(Array<T>, Builtin.RawPointer) // user: %
14
[VARIED] (%7, **%8**) = destructure_tuple %6 : $(Array<T>, Builtin.RawPointer) // user: %
9
[VARIED]   %9 = pointer_to_address %8 : $Builtin.RawPointer to [strict] $*T // users: %12
, %10
[VARIED]   %11 = integer_literal $Builtin.Word, 1          // user: %12
[VARIED]   %12 = index_addr %9 : $*T, %11 : $Builtin.Word  // user: %13
```

`%1` should be be active.

Reply:
> Activity analysis also seems not quite right:

This is actually due to a bug in the code above:
```diff
       // Propagate usefulness through `pointer_to_address` users' operands.
       for (auto use : ptai->getUses()) {
         auto *inst = use->getUser();
         if (auto *si = dyn_cast<StoreInst>(inst)) {
           setUsefulAndPropagateToOperands(si->getSrc(), dependentVariableIndex);
         } else if (auto *cai = dyn_cast<CopyAddrInst>(inst)) {
           setUsefulAndPropagateToOperands(cai->getSrc(),
                                           dependentVariableIndex);
         } else if (auto *iai = dyn_cast<IndexAddrInst>(inst)) {
           for (auto use : iai->getUses()) {
             if (auto si = dyn_cast<StoreInst>(use->getUser())) {
               setUsefulAndPropagateToOperands(si->getSrc(),
                                               dependentVariableIndex);
-            } else if (auto *cai = dyn_cast<CopyAddrInst>(inst)) {
+            } else if (auto *cai = dyn_cast<CopyAddrInst>(use->getUser())) {
               setUsefulAndPropagateToOperands(cai->getSrc(),
                                               dependentVariableIndex);
             }
           }
         }
       }
```

In light of this, and an orthogonal issue regarding non-differentiable values being marked active (TF-966), I decided to selectively propagate usefulness.